### PR TITLE
Replace gnuflag by codegangsta/cli

### PR DESCRIPTION
**Pull request by pcdummy**
This PR implements feature request #490, its a big commit but only a copy/paste and changeing the args list.

No logic changed!

Signed-off-by: Rene Jochum rene@jochums.at

### CODE REVIEW

File: lxc/action.go [hunk 1]
```diff
@@ -3,40 +3,63 @@ package main
 import (
 	"fmt"
 
+	"github.com/codegangsta/cli"
```
Comment (stgraber) [inline lxc/action.go]: We keep core, external and internal imports separated in groups throughout our codebase, as codegangsta isn't part of lxd's codebase, please add a newline after it.

File: lxc/file.go [hunk 2]
```diff
@@ -11,38 +11,93 @@ import (
 	"strings"
 	"syscall"
 
+	"github.com/codegangsta/cli"
```
Comment (stgraber) [inline lxc/file.go]: Insert a new line after this import to keep it separate from the github.com/lxc/lxd/* ones.

File: lxc/finger.go [hunk 3]
```diff
@@ -1,28 +1,23 @@
 package main
 
 import (
+	"github.com/codegangsta/cli"
```
Comment (stgraber) [inline lxc/finger.go]: Insert a new line after this import to keep it separate from the github.com/lxc/lxd/* ones.

File: lxc/copy.go [hunk 4]
```diff
@@ -4,33 +4,39 @@ import (
 	"fmt"
 	"strings"
 
+	"github.com/codegangsta/cli"
```
Comment (stgraber) [inline lxc/copy.go]: Insert a new line after this import to keep it separate from the github.com/lxc/lxd/* ones.

File: lxc/delete.go [hunk 5]
```diff
@@ -6,35 +6,42 @@ import (
 	"os"
 	"strings"
 
+	"github.com/codegangsta/cli"
```
Comment (stgraber) [inline lxc/delete.go]: Insert a new line after this import to keep it separate from the github.com/lxc/lxd/* ones.

File: lxc/info.go [hunk 6]
```diff
@@ -7,30 +7,35 @@ import (
 
 	"gopkg.in/yaml.v2"
 
+	"github.com/codegangsta/cli"
```
Comment (stgraber) [inline lxc/info.go]: Move this one to the previous block (above yaml.v2).

File: lxc/init.go [hunk 7]
```diff
@@ -2,15 +2,72 @@ package main
 
 import (
 	"fmt"
-	"os"
 	"strings"
 
+	"github.com/codegangsta/cli"
```
Comment (stgraber) [inline lxc/init.go]: Insert a new line after this import to keep it separate from the github.com/lxc/lxd/* ones.

File: lxc/launch.go [hunk 8]
```diff
@@ -4,45 +4,68 @@ import (
 	"fmt"
 	"strings"
 
+	"github.com/codegangsta/cli"
```
Comment (stgraber) [inline lxc/launch.go]: Insert a new line after this import to keep it separate from the github.com/lxc/lxd/* ones.

File: lxc/monitor.go [hunk 9]
```diff
@@ -5,11 +5,32 @@ import (
 
 	"gopkg.in/yaml.v2"
 
+	"github.com/codegangsta/cli"
```
Comment (stgraber) [inline lxc/monitor.go]: Move this one above yaml.v2

File: lxc/move.go [hunk 10]
```diff
@@ -1,33 +1,22 @@
 package main
 
 import (
+	"github.com/codegangsta/cli"
```
Comment (stgraber) [inline lxc/move.go]: Insert a new line after this import to keep it separate from the github.com/lxc/lxd/* ones.

File: lxc/profile.go [hunk 11]
```diff
@@ -9,21 +9,279 @@ import (
 
 	"gopkg.in/yaml.v2"
 
+	"github.com/codegangsta/cli"
```
Comment (stgraber) [inline lxc/profile.go]: Move this one above yaml.v2

File: lxc/publish.go [hunk 12]
```diff
@@ -4,38 +4,43 @@ import (
 	"fmt"
 	"strings"
 
+	"github.com/codegangsta/cli"
```
Comment (stgraber) [inline lxc/publish.go]: Insert a new line after this import to keep it separate from the github.com/lxc/lxd/* ones.

File: lxc/publish.go [hunk 13]
```diff
@@ -4,38 +4,43 @@ import (
...
 	"github.com/lxc/lxd"
-	"github.com/lxc/lxd/shared/gnuflag"
 	"github.com/lxc/lxd/shared/i18n"
```
Comment (stgraber) [inline lxc/publish.go]: Can you remove this new line while you are at it?

File: lxc/remote.go [hunk 14]
```diff
@@ -12,47 +12,291 @@ import (
 	"sort"
 	"strings"
 
+	"github.com/codegangsta/cli"
 	"github.com/olekukonko/tablewriter"
```
Comment (stgraber) [inline lxc/remote.go]: Can you remove this new line while you are at it?

File: lxc/restore.go [hunk 15]
```diff
@@ -3,39 +3,40 @@ package main
 import (
 	"fmt"
 
+	"github.com/codegangsta/cli"
```
Comment (stgraber) [inline lxc/restore.go]: Insert a new line after this import to keep it separate from the github.com/lxc/lxd/* ones.

File: lxc/snapshot.go [hunk 16]
```diff
@@ -3,42 +3,43 @@ package main
 import (
 	"fmt"
 
+	"github.com/codegangsta/cli"
```
Comment (stgraber) [inline lxc/snapshot.go]: Insert a new line after this import to keep it separate from the github.com/lxc/lxd/* ones.

**Comment by stgraber** [comment]
Looks good overall, having proper parser nesting is really nice!

I left a few comments from a real quick read through the big diff, they're just things that jumped at me when reading the diff and are all styling things rather than code/logic issues.

I'm hoping to have the time to do a proper review of the result of this branch over the next couple of days.

I know you said on IRC that you weren't planning for this to make it into 2.0 but if as it says, it's not causing any user visible behavioral change or regression, I don't really see a reason not too, would also mean you won't have to constantly rebase this branch until we get going with 2.1 (we expect to be doing a couple of months of bugfix only work for 2.0).

@hallyn @tych0 opinion on whether we should do the parser change for 2.0 or wait for 2.1?

**Comment by stgraber** [comment]
Oh, also, if we want to entirely kill gnuflag, lxd/main.go would also need porting to codegangsta, shouldn't be too difficult though, it doesn't use sub-parsers at all and maybe that will let us move away from the massive, hand-built "--help" we've got for it.

**Comment by stgraber** [comment]
jenkins: test this please

File: lxc/copy.go [hunk 17]
```diff
@@ -4,33 +4,39 @@ import (
...
-	ephem bool
-}
+var commandCopy = cli.Command{
+	Name:      "copy",
+	Usage:     i18n.G("Copy containers within or in between lxd instances."),
+	ArgsUsage: i18n.G("[remote:]<source container> [remote:]<destination container>"),
 
-func (c *copyCmd) showByDefault() bool {
-	return true
+	Flags: append(commandGlobalFlags, cli.BoolFlag{
```
Comment (tych0) [inline lxc/copy.go]: Is it really safe to use append like this? e.g. fedcff6dff73e4e9416c481529f67aeca09968e2

Comment (tych0) [inline lxc/copy.go]: Seems like we might want a wrapper for cli.Command that does this little append ourselves so we don't have to do it every time. And then we could do it in a safer way :)

File: lxc/launch.go [hunk 18]
```diff
@@ -4,45 +4,68 @@ import (
...
+var commandLaunch = cli.Command{
+	Name:      "launch",
+	Usage:     i18n.G("Launch a container from a particular image."),
+	ArgsUsage: i18n.G("[remote:]<image> [remote:][<name>] [--ephemeral|-e] [--profile|-p <profile>...]"),
+	Description: i18n.G(`Launches a container using the specified image and name.
+
+   Not specifying -p will result in the default profile.
+   Specifying "-p" with no argument will result in no profile.`),
+
+	Flags: []cli.Flag{
```
Comment (tych0) [inline lxc/launch.go]: Here you're forgetting the global flags if I'm not mistaken. Let's make a wrapper for this :)

File: test/suites/basic.sh [hunk 19]
```diff
@@ -77,7 +77,7 @@ test_basic_usage() {
   lxc profile create priv
   lxc profile set priv security.privileged true
   lxc init testimage barpriv -p default -p priv
-  lxc publish barpriv --alias=foo-image prop1=val1
```
Comment (tych0) [inline test/suites/basic.sh]: Does this change imply that you have to pass all `--args` before positional arguments? I always think this is annoying when I have to order things :(

Comment (tych0) [inline test/suites/basic.sh]: ...and it does indeed represent a visible change in user behavior.

**Comment by tych0** [comment]
A few comments myself, although the positional arg thing is the only big one (the `append` stuff should be fixed before this is merged, but is mostly mechanical).

**Comment by stgraber** [comment]
Jenkins failed because of:

    lxc/main_test.go:60: undefined: expandAlias

Please make sure that "sudo -E go test ./..." passes with your branch.

File: test/suites/basic.sh (hunk 19 above)
Comment (pcdummy) [inline test/suites/basic.sh]: It works either with:

    lxc publish --alias=foo-image barpiv prov1=val1

or:

    lxc publish busybox prov1=val1 --alias=busybox

**Comment by pcdummy** [comment]
lxd/* replacement with codegangsta and rebase comming on the weekend or tomorrow.

**Comment by pcdummy** [comment]
I forgot to implement the "pause" command, commit for that will come soon.

**Comment by pcdummy** [comment]
This is ready for another review. Please let me squash the commits before merge. (or squash them yourself).

File: lxc/image.go [hunk 20]
```diff
@@ -9,520 +9,654 @@ import (
...
+   metadata and rootfs tarballs).
+
+   Images can be referenced by their full hash, shortest unique partial
+   hash or alias name (if one is set).
+
+
+   lxc image import <tarball> [rootfs tarball|URL] [remote:] [--public] [--alias=ALIAS]... [prop=value]...
+   Import an image tarball (or tarballs) into the LXD image store.
+
+   lxc image copy [remote:]<image> <remote>: [--alias=ALIAS].. [--copy-aliases] [--public]
```
Comment (stgraber) [inline lxc/image.go]: missing auto-update there

**Comment by tych0** [comment]
I'd still like to raise an objection about having to order my arguments in a certain order. That's a regression from the previous behavior, it would be nice to at least look and see how easy it is to make it happen in the cli libarary.

**Comment by srkunze** [comment]
> I'd still like to raise an objection about having to order my arguments in a certain order.

I join to object.

**Comment by pcdummy** [comment]
The tests run through as i changed the argument order in the tests, i don't have the time to fix [codegangsta/cli](https://github.com/codegangsta/cli) and it has a lot of open Pull Requests not sure how long it would take to get that changed parser into it.

**Comment by techtonik** [comment]
So what is the status?

**Comment by stgraber** [comment]
The main blocker is a few missing bits in codegangsta/cli itself. Once someone gets around to introducing the bits we need, then we could unblock this and move forward with replacing the parser.

**Comment by pcdummy** [comment]
From the 8th of August i have a week of vacation, then i'll have a look into codegangsta/cli for our usecase.

**Comment by stgraber** [comment]
Cool, great to hear!

**Comment by gdouezangrard** [comment]
Told you :P #17

**Comment by pcdummy** [comment]
Haven't seen that Pull, getting codegangsta to work with LXC isn't an easy task, think we have to make lots of changes.

**Comment by techtonik** [comment]
Maybe  choose a library that is more complete? https://github.com/avelino/awesome-go#command-line

**Comment by techtonik** [comment]
@pcdummy it appears that I don't have the time, and getting port forwarding working is on top of my priority right now.

**Comment by albertodonato** [comment]
The library has now been moved to https://github.com/urfave/cli, imports should be updated

**Comment by stgraber** [comment]
Yeah, I'll close this pull request for now. I've started using urfave/cli for a number of my projects myself and will probably start switching some of the LXD commands over to it soon-ish but the port of the main client will probably be done at the same time as we do other bigger changes to it before LXD 3.0 as I'm unsure we can do the port without breaking anything.

**Comment by stgraber** [comment]
@pcdummy thanks very much for your work on this, keep the branch around as I'm sure we'll eventually pick some of that work up :)

**Pull request closed without merging**